File: tree/create_tree2.cpp
c++
// 一个树的每个节点含有一个值，父节点，子节点，请根据输入来使用C++来实现这类树的构造

// 定义一个树节点的结构体
struct TreeNode {
  int value; // 节点的值
  TreeNode* parent; // 节点的父节点
  TreeNode* left; // 节点的左子节点
  TreeNode* right; // 节点的右子节点

  // 构造函数，初始化节点的值和指针
  TreeNode(int v) {
    value = v;
    parent = nullptr;
    left = nullptr;
    right = nullptr;
  }
};

// 定义一个树的类
class Tree {
  private:
    TreeNode* root; // 树的根节点

  public:
    // 构造函数，初始化根节点为nullptr
    Tree() {
      root = nullptr;
    }

    // 根据输入来构造树
    void buildTree(vector<int>& input) {
      // 如果输入为空，返回
      if (input.empty()) return;

      // 创建一个队列，用于层次遍历
      queue<TreeNode*> q;

      // 创建一个根节点，值为输入的第一个元素，并将其入队
      root = new TreeNode(input[0]);
      q.push(root);

      // 定义一个索引变量，用于遍历输入
      int index = 1;

      // 当队列不为空时，循环以下步骤
      while (!q.empty()) {
        // 出队一个节点，作为当前节点
        TreeNode* curr = q.front();
        q.pop();

        // 如果索引没有超过输入的大小，创建一个左子节点，值为输入的当前元素，并将其入队
        if (index < input.size()) {
          TreeNode* left = new TreeNode(input[index]);
          curr->left = left;
          left->parent = curr;
          q.push(left);
          index++;
        }

        // 如果索引没有超过输入的大小，创建一个右子节点，值为输入的当前元素，并将其入队
        if (index < input.size()) {
          TreeNode* right = new TreeNode(input[index]);
          curr->right = right;
          right->parent = curr;
          q.push(right);
          index++;
        }
      }
    }

    // 返回树的根节点
    TreeNode* getRoot() {
      return root;
    }
};

File: tree/create_tree2.java
//定义TreeNode类
class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;
	//构造方法
	TreeNode(int val) {
		this.val = val;
		this.left = null;
		this.right = null;
}
}

//定义一个全局变量index，用来记录当前创建的节点的索引
int index = 0;

//定义一个方法，用来构造二叉树
TreeNode buildTree(int[] values, int[] parent) {
//如果values数组为空或者index超出范围，返回null
	if (values == null || values.length == 0 || index >= values.length) {
	return null;
}
	//创建一个新的节点，给它赋值
	TreeNode node = new TreeNode(values[index]);
	//将当前节点的索引存入parent数组
	parent[index] = -1; //假设当前节点是根节点，没有父节点
	//递增index
	index++;
	//递归地创建左子节点，并更新parent数组
	node.left = buildTree(values, parent);
	if (node.left != null) {
		parent[index - 1] = index - 2; //左子节点的父节点是当前节点
	}
	//递归地创建右子节点，并更新parent数组
	node.right = buildTree(values, parent);
	if (node.right != null) {
		parent[index - 1] = index - 2; //右子节点的父节点是当前节点
	}
	//返回当前节点作为二叉树的结果
	return node;
}
